### ISSUE TITLE
Multi-GPU distributed error

### ISSUE THREAD

**Issue by MinjeongKim03** [ultralytics/yolov5#9504]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello I run this command. But I got an error.

terminal command

python -m torch.distributed.run --nproc_per_node 2 train.py --batch 16 --img 640 --epochs 300 --data data/coco.yaml --weights yolov5s.pt --patience 300 --device 0,1  --save-period 1

error
![error](https://user-images.githubusercontent.com/94886176/191148318-0ffef1bd-3da9-4b50-9a21-11ae02fd6ff5.png)

**Comment by glenn-jocher** [comment]
@MinjeongKim03 it appears you may have environment problems. Please ensure you meet all dependency requirements if you are attempting to run YOLOv5 locally.  If in doubt, create a new virtual [Python 3.9](https://www.python.org/) environment, clone the latest repo ([code changes daily](https://github.com/ultralytics/yolov5/commits/master)), and `pip install` [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) again from scratch.

💡 ProTip! Try one of our verified environments below if you are having trouble with your local environment.


## Requirements

[**Python>=3.7.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.7**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

[Models](https://github.com/ultralytics/yolov5/tree/master/models) and [datasets](https://github.com/ultralytics/yolov5/tree/master/data) download automatically from the latest YOLOv5 [release](https://github.com/ultralytics/yolov5/releases) when first requested.

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), validation ([val.py](https://github.com/ultralytics/yolov5/blob/master/val.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/export.py)) on MacOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by MinjeongKim03** [comment]
I changed to Python 3.9 environment and tried again, but the following error occurred.

Traceback (most recent call last):                                                                                        File "/home/()/()/yolov5/train.py", line 630, in <module>                                                   Traceback (most recent call last):                                                                                        File "/home/()/()/yolov5/train.py", line 630, in <module>                                                       main(opt)                                                                                                                 File "/home/()/()/yolov5/train.py", line 526, in main                                                   main(opt)                                                                                                                 File "/home/()/()/yolov5/train.py", line 526, in main                                                           train(opt.hyp, opt, device, callbacks)                                                                                File "/home/()/()/yolov5/train.py", line 109, in train                                                          train(opt.hyp, opt, device, callbacks)                                                                                File "/home/()/()/yolov5/train.py", line 110, in train                                                          with torch_distributed_zero_first(LOCAL_RANK):                                                                        File "/home/()/.conda/envs/()/lib/python3.9/contextlib.py", line 119, in __enter__                            data_dict = data_dict or check_dataset(data)  # check if None                                                         File "/home/()/.conda/envs/()/lib/python3.9/contextlib.py", line 126, in __exit__                             return next(self.gen)                                                                                                 File "/home/()/()/yolov5/utils/torch_utils.py", line 91, in torch_distributed_zero_first                        next(self.gen)                                                                                                        File "/home/()/()/yolov5/utils/torch_utils.py", line 94, in torch_distributed_zero_first                        dist.barrier(device_ids=[local_rank])                                                                                 File "/home/qisens/.conda/envs/minjeong_2/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 2784, in barrier                                                                                                               dist.barrier(device_ids=[0])                                                                                          File "/home/()/.conda/envs/()/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 2784, in barrier                                                                                                                   work = default_pg.barrier(opts=opts)work = default_pg.barrier(opts=opts)                                                                                                                                                                RuntimeErrorRuntimeError: : NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:46, unhandled cuda error, NCCL version 2.10.3                                                                                                    ncclUnhandledCudaError: Call to CUDA function failed.NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:46, unhandled cuda error, NCCL version 2.10.3                                                                           ncclUnhandledCudaError: Call to CUDA function failed.                                                                                                                                                                                           ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 0 (pid: 2416573) of binary: /home/()/.conda/envs/()/bin/python                                                                                 Traceback (most recent call last):                                                                                        File "/home/()/.conda/envs/()/lib/python3.9/runpy.py", line 197, in _run_module_as_main                       return _run_code(code, main_globals, None,                                                                            File "/home/()/.conda/envs/()/lib/python3.9/runpy.py", line 87, in _run_code                                  exec(code, run_globals)                                                                                               File "/home/()/.conda/envs/()/lib/python3.9/site-packages/torch/distributed/run.py", line 765, in <module>    main()                                                                                                                File "/home/()/.conda/envs/()/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 345, in wrapper                                                                                         return f(*args, **kwargs)                                                                                             File "/home/()/.conda/envs/()/lib/python3.9/site-packages/torch/distributed/run.py", line 761, in main        run(args)                                                                                                             File "/home/()/.conda/envs/()/lib/python3.9/site-packages/torch/distributed/run.py", line 752, in run         elastic_launch(                                                                                                       File "/home/()/.conda/envs/()/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 131, in __call__                                                                                                                   return launch_agent(self._config, self._entrypoint, list(args))                                                       File "/home/()/.conda/envs/()/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 245, in launch_agent                                                                                                               raise ChildFailedError(                                                                                             torch.distributed.elastic.multiprocessing.errors.ChildFailedError:                                                      ============================================================                                                            train.py FAILED                                                                                                         ------------------------------------------------------------                                                            Failures:                                                                                                               [1]:                                                                                                                      time      : 2022-09-21_13:56:16                                                                                         host      :()                                                                                         rank      : 1 (local_rank: 1)                                                                                           exitcode  : 1 (pid: 2416574)                                                                                            error_file: <N/A>                                                                                                       traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html                              ------------------------------------------------------------                                                            Root Cause (first observed failure):                                                                                    [0]:                                                                                                                      time      : 2022-09-21_13:56:16                                                                                         host      : ()                                                                                         rank      : 0 (local_rank: 0)                                                                                           exitcode  : 1 (pid: 2416573)                                                                                            error_file: <N/A>                                                                                                       traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html                              ============================================================

**Comment by glenn-jocher** [comment]
@MinjeongKim03 I would run DDP training in our Docker image if you are having local environment issues:

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

**Comment by gra-ch** [comment]
I tried to use Docker image but still got the same error.

**Comment by glenn-jocher** [comment]
@gra-ch I just tried the DDP usage example shown in train.py in our Docker image and discovered a bug related to ClearML, so I've removed it from the Dockerfile requirements in https://github.com/ultralytics/yolov5/pull/9876

**Comment by glenn-jocher** [comment]
Good news 😃! Docker training issue related to ClearML should now be fixed ✅ in PR #9876. 

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
